synapse/lib/lmdb.py

```diff
@@ -0,0 +1,314 @@
+
+                name = lkey.decode('utf8')
+                valu = struct.unpack('>Q', lval)[0]
+
+                self.info[name] = valu
+
+            indx = 0
+
+            curs = xact.cursor(db=self._db_history)
+            if curs.last():
+                indx = struct.unpack('>Q', curs.key())[0]
+
+            self.indx = itertools.count(indx)
+
+    def inc(self, xact, prop, step=1):
+        '''
+        Increment the value of a global metric.
+        '''
+        valu = self.info.get(prop, 0)
+        valu += step
+
+        self.info[prop] = valu
+
+        penc = prop.encode('utf8')
+        pval = struct.pack('>Q', valu)
+
+        xact.put(penc, pval, db=self._db_current)
+
+    def stat(self):
+        return self.info
+
+    def iter(self, xact, offs):
+
+        lkey = struct.pack('>Q', offs)
+
+        with xact.cursor(db=self._db_history) as curs:
+
+            if not curs.set_key(lkey):
+                return
+
+            for lkey, lval in curs.iternext():
+
+                indx = struct.unpack('>Q', lkey)[0]
+                sample = s_msgpack.un(lval)
+
+                yield indx, sample
+
+    def record(self, xact, info):
+        '''
+        Record metrics info.
+
+        Args:
+            xact (Transaction): An LMDB write transaction.
+            info (dict): A dictionary of sample info to save.
+        '''
+        indx = struct.pack('>Q', next(self.indx))
+        sample = s_msgpack.en(info)
+        xact.put(indx, sample, db=self._db_history, append=True)
+
+class PropSetr:
+    '''
+    A helper for setting properties.  Most to cache cursors.
+    '''
+    def __init__(self, psto, xact):
+
+        self.xact = xact
+        self.psto = psto
+
+        self.purs = xact.cursor(db=psto.props)
+        self.burs = xact.cursor(db=psto.byprop)
+
+    def has(self, penc, byts):
+        return self.burs.set_key(penc + b'\x00' + byts)
+
+    #def rem(self, buid):
+    #def addtag(self, buid, tag, times):
+    #def deltag(self, buid, tag):
+
+    def set(self, buid, penc, lval, flags=0):
+
+        pkey = buid + penc
+
+        noindex = flags & STOR_FLAG_NOINDEX
+        multival = flags & STOR_FLAG_MULTIVAL
+
+        if self.purs.set_key(pkey):
+
+            # At a minimum, we have the key...
+            if flags & STOR_FLAG_DEFVAL:
+                return False
+
+            # if this is a multi-val, iter to check for valu
+            if multival:
+
+                # if a multival prop=valu exists, skip adding
+                for mval in self.purs.iternext_dup():
+                    if mval == lval:
+                        return False
+
+            else:
+
+                # if it's exactly what we want, skip...
+                oldb = self.purs.value()
+                if oldb == lval:
+                    return False
+
+                okey = penc + b'\x00' + oldb
+                if self.burs.set_key(okey):
+                    self.burs.delete()
```

Comment:
`Cursor.delete` doesn't take key,val args and instead just deletes the current item. Is this the right way to go about deleting here? `set_key` seeks to the key and returns True if it was found.

Reply:
bump